match/lib/match/storage.rb

```diff
@@ -3,13 +3,35 @@
 
 module Match
   module Storage
-    def self.for_mode(storage_mode, params)
-      if storage_mode == "git"
-        return Storage::GitStorage.configure(params)
-      elsif storage_mode == "google_cloud"
-        # return Storage::GoogleCloudStorage
-      else
-        UI.user_error!("Invalid storage mode '#{storage_mode}'")
+    class << self
+      def backends
+        @backends ||= {
+          "git" => lambda { |params|
+            return Storage::GitStorage.configure(params)
+          }
+        }
+      end
+
+      def register_backend(type: nil, storage_class: nil, &configurator)
+        UI.user_error!("No type specified for storage backend") if type.nil?
+
+        normalized_name = type.to_s
+        UI.message("Replacing Match::Encryption backend for type '#{normalized_name}'") if backends.include?(normalized_name)
+
+        if configurator
+          @backends[normalized_name] = configurator
+        elsif storage_class
+          @backends[normalized_name] = ->(params) { return storage_class.configure(params) }
+        else
+          UI.user_error!("Specify either a `storage_class` or a configuration block when registering a storage backend")
+        end
+      end
+
+      def for_mode(storage_mode, params)
```

Comment:
The names `for_mode` and `for_storage_mode` were taken over from how they were named previously

I thought about refactoring this into a module, but thought it might make what's happening here non-obvious, so going by principle of least surprise decided to violate DRY. If you have a better way or think that having this refactored out would make more sense, I'm open to changing it

Reply:
Sounds like a good reason to me, makes sense 👍:)